core/src/main/java/com/linecorp/armeria/internal/common/GracefulConnectionShutdownHandler.java

```diff
@@ -60,29 +60,40 @@ protected GracefulConnectionShutdownHandler(long drainDurationMicros) {
     /**
      * Code executed on the connection drain end. Executed at most once.
      */
-    protected abstract void onDrainEnd(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception;
+    protected abstract void onDrainEnd(ChannelHandlerContext ctx) throws Exception;
 
     public void start(ChannelHandlerContext ctx, ChannelPromise promise) {
-        if (this.promise == null) {
-            this.promise = promise;
-        } else {
-            // Chain promises in case start is called multiple times.
-            this.promise.addListener((ChannelFutureListener) future -> {
-                if (future.cause() != null) {
-                    promise.setFailure(future.cause());
-                } else {
-                    promise.setSuccess();
-                }
-            });
+        final Channel ch = ctx.channel();
+        if (!ch.isActive()) {
+            promise.trySuccess();
+            return;
         }
-        run(ctx, this.promise);
+
+        started = true;
+
+        // Ensure the given promise is fulfilled when the channel is closed.
+        ch.closeFuture().addListener((ChannelFutureListener) future -> {
+            if (future.cause() != null) {
+                promise.tryFailure(future.cause());
+            } else {
+                promise.trySuccess();
+            }
+        });
+
+        run(ctx);
     }
 
-    private void run(ChannelHandlerContext ctx, ChannelPromise promise) {
+    private void run(ChannelHandlerContext ctx) {
         if (drainFuture != null) {
+            final boolean cancelled;
             if (drainFuture.getDelay(TimeUnit.MICROSECONDS) > drainDurationMicros) {
                 // Maybe reschedule below.
-                drainFuture.cancel(false);
+                cancelled = drainFuture.cancel(false);
+            } else {
+                cancelled = false;
```

Comment:
nit:
```suggestion
                return;
```

Reply:
It's intentional because otherwise I'll have to duplicate the comment below. :smile:

Reply:
SGTM. 😄